File: server-shared/src/main/java/eu/darken/porter/core/ServerPolicy.java
package eu.darken.porter.core;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import rikka.shizuku.server.ClientRecord;

/**
 * The decisions a server makes for itself. Every method has a default, so a server that grants
 * nothing is a valid one and a test implements only the hooks it drives.
 */
public interface ServerPolicy {

    default boolean checkCallerPermission(
            @NonNull String func, @NonNull CallerIdentity caller, @Nullable ClientRecord record) {
        return false;
    }

    default boolean checkCallerManagerPermission(@NonNull String func, @NonNull CallerIdentity caller) {
        return false;
    }

    default void showPermissionConfirmation(
            int requestCode, @NonNull ClientRecord record, @NonNull CallerIdentity caller, int userId) {
        throw new UnsupportedOperationException("no permission confirmation UI");
    }

    /** Before a record exists for the caller. */
    default void onAttaching(@NonNull CallerIdentity caller, @NonNull String packageName) {
    }

    /** The reply the endpoint is about to send, still open to shaping. */
    default void onAttached(@NonNull ClientRecord record, boolean created, @NonNull Bundle reply) {
    }

    /** After the reply reached the client. */
    default void onBound(@NonNull ClientRecord record, boolean created) {
    }
}

File: server-shared/src/main/java/eu/darken/porter/core/ServerPolicy.kt
package eu.darken.porter.core

import android.os.Bundle
import rikka.shizuku.server.ClientRecord

/**
 * The decisions a server makes for itself. Every method has a default, so a server that grants
 * nothing is a valid one and a test implements only the hooks it drives.
 */
interface ServerPolicy {

    fun checkCallerPermission(func: String, caller: CallerIdentity, record: ClientRecord?): Boolean = false

    fun checkCallerManagerPermission(func: String, caller: CallerIdentity): Boolean = false

    fun showPermissionConfirmation(requestCode: Int, record: ClientRecord, caller: CallerIdentity, userId: Int) {
        throw UnsupportedOperationException("no permission confirmation UI")
    }

    /** Before a record exists for the caller. */
    fun onAttaching(caller: CallerIdentity, packageName: String) {
    }

    /** The reply the endpoint is about to send, still open to shaping. */
    fun onAttached(record: ClientRecord, created: Boolean, reply: Bundle) {
    }

    /** After the reply reached the client. */
    fun onBound(record: ClientRecord, created: Boolean) {
    }
}
